### PR TITLE
GSFA example

### PR DESCRIPTION
**Pull request by nkgevorgyan**

### CODE REVIEW

File: source/tutorial/gsfa/GSFA.rst [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+
+	>>> %matplotlib inline
+	>>> import matplotlib.pyplot as plt
+	>>>
+	>>> from sklearn import (datasets, model_selection, metrics)
+	>>> import numpy as np
+	>>>
+	>>> import mdp 
+	>>> from sklearn.svm import SVC
+
```
Comment (Stewori) [inline source/tutorial/gsfa/GSFA.rst]: You are using plenty of blank lines, maybe too many? Please look into other rst files in this repository, e.g. https://raw.githubusercontent.com/mdp-toolkit/mdp-docs/master/source/tutorial/nodes.rst and try to adopt the style. Two consecutive blank lines are rare there, only at the end of high-level sections. I think three consecutive blank lines never occur, but I saw this in your rst (somewhere below).

File: source/tutorial/gsfa/GSFA.rst [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+- WDBC-Malignant
+- WDBC-Benign
+
+
+First let’s import the usual data science modules and library
+`mdp <https://mdp-toolkit.github.io>`__. We use it for GSFA
+implementation and other data processing tools.
+
+
+	>>> %matplotlib inline
```
Comment (Stewori) [inline source/tutorial/gsfa/GSFA.rst]: I think this is ipython notebook specific and should be removed in the rst

Comment (Stewori) [reply source/tutorial/gsfa/GSFA.rst]: (please apply to other occurrences as well if any)

File: source/tutorial/gsfa/GSFA.rst [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+:math:`G = (\textbf{V}, \textbf{E})` *(as illustrated in Figure bellow)*
+s.t. :
+
+-  :math:`\textbf{V}` corresponds to set of :math:`\textbf{x}(n)`
+   :math:`\rightarrow` *each vertex of the graph is a sample*
+
+-  Each edge of :math:`\textbf{E}` corresponds to a pair of samples
+   :math:`(\textbf{x}(n), \textbf{x}(n'))`
+
+|
```
Comment (Stewori) [inline source/tutorial/gsfa/GSFA.rst]: Is this by intention? Can it be cleaned up? I.e. the blank lines with leading '|'

**Comment by NiMlr** [comment]
When compiling the proposed version of the docs I found the following new warnings.

```
./mdp-docs/source/tutorial/gsfa/GSFA.rst: WARNING: document isn't included in any toctree
./mdp-docs/source/tutorial/gsfa/GSFA.rst:268: WARNING: Explicit markup ends without a blank line; unexpected unindent.
./mdp-docs/source/tutorial/gsfa/GSFA.rst:390: WARNING: Block quote ends without a blank line; unexpected unindent.
./mdp-docs/source/tutorial/gsfa/GSFA.rst:396: (SEVERE/4) Title level inconsistent
```

As there are a lot of old accumulated warnings I know it is painful to work with. A way to make it easier is to disable parts of the docs such anything that parses the MDP-toolkit source code, as this should have the most warnings. Please limit these changes to your local copy that helps you create the tutorial.

**Comment by NiMlr** [comment]
If I am not mistaken, by context of the [current MDP-Documentation](https://mdp-toolkit.github.io/documentation.html) you are providing a example instead of a tutorial. Please compare with it and change the terminology and file paths accordingly.

File: source/tutorial/gsfa/GSFA.rst [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+edges represent similarities of the corresponding labels. Later, we use
+aquired low-dimensional representaion of the original data to train
+typical supervised learning algortihms.
+
+In this tutorial we will briefly exlpain the idea behind GSFA and
+specify the optimization task it solves. Moreover, we will show the
+effeciency of the GSFA compared to SVM on a toy example and introduce an
+approach that makes classsification task more interpretable.
+
+Table of Contents
```
Comment (NiMlr) [inline source/tutorial/gsfa/GSFA.rst]: Can this be listed as a [Sphinx Sub-TOC](https://www.sphinx-doc.org/en/1.5/markup/toctree.html)? Also please add the file to the TOC-tree in the example section.

File: source/tutorial/gsfa/GSFA.rst [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+
+2\. Each vertex  :math:`\textbf{x(}n\textbf{)}` has an associated weight :math:`v_n > 0`, which can be used to reflect its importance, frequency, or reliability.
+
+For instance, a sample occurring frequently in an observed phenomenon
+should have a larger weight than a rare sample. This representation
+includes the standard time series as a special case in which the graph
+has a linear structure and all node and edge weights are identical *(as
+illustrated in Figure(b))*.
+
+.. image:: training_graph.png
```
Comment (NiMlr) [inline source/tutorial/gsfa/GSFA.rst]: Some images appear not to be centered correctly and hard to read. Maybe there is a way to change this. If you need some help in plotting, let me know.

File: source/tutorial/gsfa/GSFA.rst [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+.. _GSFA Optimization Problem:
+
+4. GSFA optimization problem 
+----------------------------
+
+The GSFA optimization problem over :math:`N` training samples can be
+stated as follows.
+
+
+-  **Given**: :math:`I` - dimensional input
```
Comment (NiMlr) [inline source/tutorial/gsfa/GSFA.rst]: Maybe there is a nicer way to include Pseudo-Code (based on html/css). Although, to me this has lower priority.

File: source/tutorial/gsfa/GSFA.rst [hunk 7]
```diff
@@ -0,0 +1,414 @@
...
+| 
+
+  .. math:: y = W^T (\textbf{x}(n) -  \hat{\textbf{x}}) \hspace{0.5cm}
+| where
+  :math:`\hspace{0.5cm} \Delta(y_j) = \lambda_j \hspace{0.5cm} 1 \leq j \leq J`
+
+--------------
+
+
+Referencies
```
Comment (NiMlr) [inline source/tutorial/gsfa/GSFA.rst]: Please check out the [section of footnotes and citations in the Sphinx-Docs](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#footnotes).

File: source/conf.py [hunk 8]
```diff
@@ -233,3 +235,6 @@ def autodoc_process_docstring(app, what, name, obj, options, lines):
     app.connect('autodoc-process-signature', autodoc_process_signature)
     app.connect('autodoc-skip-member', autodoc_skip_member)
     app.connect('autodoc-process-docstring', autodoc_process_docstring)
+
+# link that is used to download mathjax
```
Comment (NiMlr) [inline source/conf.py]: Is this essential? On first glance, the result appeared quite similar when dropping it.

**Review by NiMlr — COMMENTED** [review]
Thanks a lot for the contribution! Please check out my comments.

**Comment by Stewori** [comment]
@NiMlr Do you think it would make sense to apply ``.. codesnippet::`` in this document, e.g. like in https://raw.githubusercontent.com/mdp-toolkit/mdp-docs/master/source/tutorial/nodes.rst?

**Comment by Stewori** [comment]
@nkgevorgyan Beside all these tiny technical issues I would like to express my appreciation for this example. It turned out really well.

**Comment by NiMlr** [comment]
> @NiMlr Do you think it would make sense to apply `.. codesnippet::` in this document, e.g. like in https://raw.githubusercontent.com/mdp-toolkit/mdp-docs/master/source/tutorial/nodes.rst?

Good point! I have never made use of it. But from what I can tell, that is precisely its purpose. Sorry, I missed that in the review.

File: source/tutorial/gsfa/GSFA.rst [hunk 9]
```diff
@@ -0,0 +1,414 @@
...
+	>>> import numpy as np
+	>>> import mdp 
+	>>> from sklearn.svm import SVC
+
+
+Now from the sklearn let’s load the ‘breast_cancer’ dataset. We’ll use
+20% of the data for testing.
+
+
+	>>> data, label = datasets.load_breast_cancer(return_X_y=True
```
Comment (NiMlr) [inline source/tutorial/gsfa/GSFA.rst]: You are possibly missing a parenthesis here.

File: source/tutorial/gsfa/GSFA.rst [hunk 10]
```diff
@@ -0,0 +1,414 @@
...
+	>>> data_train, data_test, label_train, label_test = \
+	... model_selection.train_test_split(data, label, test_size=0.2)
+
+We get the baseline classification quality.
+
+	>>> clf = SVC(gamma='auto')
+	>>> clf.fit(data_train, label_train)
```
Comment (NiMlr) [inline source/tutorial/gsfa/GSFA.rst]: This and the previous line is also run in 124 + 125. Is both needed?

File: source/tutorial/gsfa/GSFA.rst [hunk 11]
```diff
@@ -0,0 +1,414 @@
...
+	>>> SVM_test = clf.predict(data_train)
+	>>> print("SVM test score: ", metrics.accuracy_score(label_train, SVM_test))
```
Comment (NiMlr) [inline source/tutorial/gsfa/GSFA.rst]: This should probably say training score.

File: source/tutorial/gsfa/GSFA.rst [hunk 12]
```diff
@@ -0,0 +1,414 @@
...
+**SVM train score:  1.0**
+
+
+	>>> clf = SVC(gamma='auto')
+	>>> clf.fit(data_train, label_train)
+	>>> SVM_test = clf.predict(data_test)
+	>>> print("SVM test score: ", metrics.accuracy_score(label_test, SVM_test))
+
+
+**SVM test score:  0.5614**
```
Comment (NiMlr) [inline source/tutorial/gsfa/GSFA.rst]: I get 0.5877192982456141 running this code.

Comment (NiMlr) [reply source/tutorial/gsfa/GSFA.rst]: You can add the result [like this](https://github.com/mdp-toolkit/mdp-docs/blame/master/source/examples/logmap/logmap.rst#L106) into the code.

File: source/tutorial/gsfa/GSFA.rst [hunk 13]
```diff
@@ -0,0 +1,414 @@
...
+	>>> f = plt.figure(figsize=(15,6))
+	>>> ax = f.add_subplot(121)
+	>>> ax2 = f.add_subplot(122)
+	>>> ax.scatter(GSFA_train[:, 0], GSFA_train[:, 1], c=label_train,
+	...	cmap=matplotlib.colors.ListedColormap(colors))
+	>>> ax2.scatter(GSFA_test[:, 0], GSFA_test[:, 1], c=label_test,
+	...	cmap=matplotlib.colors.ListedColormap(colors))
+	>>> ax.set_title("Train data in 2-D")
+	>>> ax2.set_title("Test data in 2-D")
+
```
Comment (NiMlr) [inline source/tutorial/gsfa/GSFA.rst]: @Stewori Should there be a `plt.show()` here?

**Review by NiMlr — COMMENTED** [review]
Here are some minor comments.

File: source/tutorial/gsfa/GSFA.rst (hunk 12 above)
Comment (NiMlr) [reply source/tutorial/gsfa/GSFA.rst]: The result is random here. If you use the codesnippets extension it will add the following lines automatically to the generated source file. I am not sure whether this can or should be changed. In any case this fixes the seed and produces consistent results.

```
import mdp
import numpy as np
np.random.seed(0)
```

File: source/tutorial/gsfa/GSFA.rst (hunk 13 above)
Comment (Stewori) [reply source/tutorial/gsfa/GSFA.rst]: Yes, @nkgevorgyan please add this here. I guess this was forgotten due to porting the code from an ipython notebook.

File: source/examples/examples.rst [hunk 14]
```diff
@@ -12,6 +12,7 @@ Examples
    gng/gng.rst
    convolution/image_convolution.rst
    scikits_learn/digit_classification.rst
+   gsfa/gsfa.rst
```
Comment (NiMlr) [inline source/examples/examples.rst:15]: Compiling the docs I get some new warnings. One of them is the following:

```sh
./mdp-docs/source/examples/examples.rst:6: WARNING: toctree contains reference to nonexisting document u'examples/gsfa/gsfa'
```
Your filename is different to the one referenced here and should be lowercase letters I think. You can check whether there are new warning by text searching "gsfa" with your command line tool after compiling the docs.

File: source/examples/gsfa/GSFA.rst [hunk 15]
```diff
@@ -0,0 +1,412 @@
...
+Then we find projections of the train and test data on feature space we have found.
+    
+	>>> GSFA_train = GSFA_n.execute(data_train)
+	>>> GSFA_test = GSFA_n.execute(data_test)
+
+Since we consider dimensionality reduction to the value of 2 we can
+try to depict train and test data on 2-D graph.
+
+
+	>>> import matplotlib
```
Comment (NiMlr) [inline source/examples/gsfa/GSFA.rst]: This should probably be moved to the top of the file.

File: source/examples/gsfa/GSFA.rst [hunk 16]
```diff
@@ -0,0 +1,412 @@
...
+:math:`\hspace{0.5cm} \Delta(y_j) = \lambda_j \hspace{0.5cm} 1 \leq j \leq J`
+
+--------------
+
+
+Referencies
+------------
+
+
+.. [1] Escalante-B. et al. (2013) `How to Solve Classification and Regression Problems on High-Dimensional Data with a Supervised Extension of Slow Feature Analysis <https://jmlr.csail.mit.edu/papers/v14/escalante13a.html>`__
```
Comment (NiMlr) [inline source/examples/gsfa/GSFA.rst]: The references create new warnings when compiling the docs, as there are no inline references referring to [1], [2], ...
Maybe you can add them in the text above to avoid the warnings. I am not sure if there is a functionality in rst that does not require this.

File: source/examples/examples.rst [hunk 17]
```diff
@@ -35,6 +37,8 @@ applications:
   of caching extension.
 * :ref:`digit_classification` — Use the combined power of MDP and scikits.learn
```
Comment (NiMlr) [inline source/examples/examples.rst:38]: There may be an inconsistency of "— " with "-" two lines below.

File: source/examples/gsfa/GSFA.rst [hunk 18]
```diff
@@ -0,0 +1,412 @@
...
+	>>> ax2 = f.add_subplot(122)
+	>>> ax.scatter(GSFA_train[:, 0], GSFA_train[:, 1], c=label_train,
+	...	cmap=matplotlib.colors.ListedColormap(colors))
+	>>> ax2.scatter(GSFA_test[:, 0], GSFA_test[:, 1], c=label_test,
+	...	cmap=matplotlib.colors.ListedColormap(colors))
+	>>> ax.set_title("Train data in 2-D")
+	>>> ax2.set_title("Test data in 2-D")
+	>>> plt.show()
+
+
```
Comment (NiMlr) [inline source/examples/gsfa/GSFA.rst]: There should be a margin here ;)

File: source/examples/gsfa/GSFA.rst [hunk 19]
```diff
@@ -0,0 +1,412 @@
+.. _gsfa:
+
+======================
+Graph-Based SFA (GSFA)
+======================
+.. codesnippet::
+
+-  Extension of slow feature analysis
```
Comment (NiMlr) [inline source/examples/gsfa/GSFA.rst]: Please capitalize the word's first letters the name of algorithms (there are other instances of this in the example). Also consider making not only the abbreviation but the name+abbreviation a link.

For headings see: [Heading Capitalization Tool](https://capitalizemytitle.com/).

File: source/examples/gsfa/GSFA.rst [hunk 20]
```diff
@@ -13,32 +14,21 @@ Graph-Based SFA (GSFA)
...
-3. `How the graphs are used for the training`_
-
-4. `GSFA Optimization Problem`_
-
-5. `Linear GSFA Algorithm`_
 
+.. contents:: **Table of contents**
+   :local:
 
 --------------
```
Comment (NiMlr) [inline source/examples/gsfa/GSFA.rst]: Is there a reason this sequence of hyphens is shorter than the one 5 lines below?

File: source/examples/gsfa/GSFA.rst [hunk 21]
```diff
@@ -377,38 +365,31 @@ and :math:`\Lambda` is a diagonal matrix with diagonal elements
...
+where
+
+.. math:: \hspace{0.5cm} \Delta(y_j) = \lambda_j \hspace{0.5cm} 1 \leq j \leq J
 
 --------------
 
 
-Referencies 
-===============
+Referencies
```
Comment (NiMlr) [inline source/examples/gsfa/GSFA.rst]: Please revise the spelling in the documents :)

**Review by NiMlr — APPROVED** [review]

File: source/examples/gsfa/GSFA.rst (hunk 19 above)
Comment (Stewori) [reply source/examples/gsfa/GSFA.rst]: And make sure the heading is capitalized the same way as in the TOC. Currently there are differences between capitalization of the headings and the TOC.

File: source/examples/gsfa/GSFA.rst [hunk 22]
```diff
@@ -0,0 +1,412 @@
...
+SFA algorithm that relies on particular graphs structure to extract
+features that preserve label similarities. More precisely, the algorithm
+utilize training graphs, in which the verticies are the samples and the
+edges represent similarities of the corresponding labels. Later, we use
+aquired low-dimensional representaion of the original data to train
+typical supervised learning algortihms.
+
+In this example we will briefly exlpain the idea behind GSFA and
+specify the optimization task it solves. Moreover, we will show the
+effeciency of the GSFA compared to SVM on a toy dataset and introduce an
```
Comment (Stewori) [inline source/examples/gsfa/GSFA.rst]: I know, grammerly suggested to add "the" before GSFA here (or was it somewhere else?). In case of GSFA, however, grammerly mistakes it to be a thing rather than a name. Since it actually is a name, there should be no "the", unless you write "the GSFA algorithm" or something. On the other hand I feel like there should be a "the" before "SVM" because the "M" stands for "machine" which is a thing. Certainly, in the line below it should either be "the classification task" or "classification tasks" (in plural but without "the").

File: source/examples/gsfa/GSFA.rst (hunk 20 above)
Comment (nkgevorgyan) [reply source/examples/gsfa/GSFA.rst]: The sequence on line 33 is used as a dividing line, while the one on line 38 is a heading.

File: source/examples/examples.rst (hunk 17 above)
Comment (nkgevorgyan) [reply source/examples/examples.rst:38]: "__" (double underscore sign) is used for some pages to refer to the links below, while "-" to refer to RST pages within the documentation.

And there is actually a vertical space between these two groups of links.

File: source/examples/examples.rst [hunk 23]
```diff
@@ -35,6 +37,8 @@ applications:
...
   in an applciation for handwritten digit classification
+* :ref:`gsfa` - Use graph-based Slow Feature Analysis (GSFA) for supervised
```
Comment (NiMlr) [inline source/examples/examples.rst]: Here we have `-`. All other examples use `—`.

**Comment by Stewori** [comment]
@NiMlr The changes look good to me. Thanks!
